context-propagation/src/main/java/io/micrometer/context/DefaultContextSnapshotFactory.java

```diff
@@ -0,0 +1,194 @@
+/**
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.context;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Predicate;
+
+class DefaultContextSnapshotFactory implements ContextSnapshotFactory {
+
+    private static final DefaultContextSnapshot emptyContextSnapshot = new DefaultContextSnapshot(new ContextRegistry(),
+            false);
+
+    private static final DefaultContextSnapshot clearingEmptyContextSnapshot = new DefaultContextSnapshot(
+            new ContextRegistry(), true);
+
+    private final ContextRegistry contextRegistry;
+
+    private final boolean clearMissing;
+
+    private final Predicate<Object> captureKeyPredicate;
+
+    public DefaultContextSnapshotFactory(ContextRegistry contextRegistry, boolean clearMissing,
+            Predicate<Object> captureKeyPredicate) {
+        this.contextRegistry = contextRegistry;
+        this.clearMissing = clearMissing;
+        this.captureKeyPredicate = captureKeyPredicate;
+    }
+
+    @Override
+    public ContextSnapshot captureAll(Object... contexts) {
+        return captureAll(contextRegistry, captureKeyPredicate, clearMissing, contexts);
+    }
+
+    @Override
+    public ContextSnapshot captureFrom(Object... contexts) {
+        return captureFromContext(captureKeyPredicate, clearMissing, contextRegistry, null, contexts);
+    }
+
+    @Override
+    public <C> ContextSnapshot.Scope setThreadLocalsFrom(Object sourceContext, String... keys) {
+        if (keys == null || keys.length == 0) {
+            return setAllThreadLocalsFrom(sourceContext, contextRegistry, clearMissing);
+        }
+        else {
+            return setThreadLocalsFrom(sourceContext, contextRegistry, clearMissing, keys);
+        }
+    }
```

Comment:
I think it would make sense for this method to move further down after the `capture*` methods, so that capture methods are together, and `setThreadLocals*` methods are together.

Reply:
I did a few more rearrangements.